File: api/app/schemas/source.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import Field, HttpUrl, model_validator

from app.schemas import AppBaseModel


class SourceCreate(AppBaseModel):
    """POST /sources request body."""

    name: str = Field(..., min_length=1, max_length=100)
    rss_url: HttpUrl
    site_url: Optional[HttpUrl] = None
    category: Optional[str] = Field(None, max_length=50)


class SourceUpdate(AppBaseModel):
    """PUT /sources/{id} request body.

    全フィールド Optional で部分更新をサポート。
    exclude_unset=True で未送信フィールドは変更しない。
    rss_url/name は明示的 null 送信を禁止（DB NOT NULL 制約保護）。
    """

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    rss_url: Optional[HttpUrl] = None
    site_url: Optional[HttpUrl] = None
    category: Optional[str] = Field(None, max_length=50)
    is_active: Optional[bool] = None

    @model_validator(mode="after")
    def reject_null_required_fields(self) -> "SourceUpdate":
        """rss_url, name が明示的に null で送信された場合を拒否する。

        exclude_unset=True では「未送信」と「null送信」を区別できるが、
        null が set された場合に DB の NOT NULL 制約で 500 になるのを防ぐ。
        model_fields_set で「送信されたが null」を検出する。
        """
        if "rss_url" in self.model_fields_set and self.rss_url is None:
            raise ValueError("rss_url cannot be null")
        if "name" in self.model_fields_set and self.name is None:
            raise ValueError("name cannot be null")
        return self


class SourceResponse(AppBaseModel):
    """GET /sources, GET /sources/{id} response."""

    id: uuid.UUID
    name: Optional[str] = None
    rss_url: str
    site_url: Optional[str] = None
    category: Optional[str] = None
    is_active: bool
    created_at: datetime


class SourceListResponse(AppBaseModel):
    """GET /sources paginated response."""

    items: list[SourceResponse]
    total: int
    page: int
    per_page: int

File: api/src/schemas/source.ts
import { z } from "zod";

export const sourceCreateSchema = z
  .object({
    name: z.string().min(1).max(100),
    rss_url: z.string().url(),
    site_url: z.string().url().nullish(),
    category: z.string().max(50).nullish(),
  })
  .strict();

export type SourceCreate = z.infer<typeof sourceCreateSchema>;

export const sourceUpdateSchema = z
  .object({
    name: z.string().min(1).max(100).nullish(),
    rss_url: z.string().url().nullish(),
    site_url: z.string().url().nullish(),
    category: z.string().max(50).nullish(),
    is_active: z.boolean().nullish(),
  })
  .strict()
  .refine(
    (data) => {
      // Reject explicit null on required fields (DB NOT NULL constraint protection)
      if (data.name === null) return false;
      if (data.rss_url === null) return false;
      return true;
    },
    { message: "name and rss_url cannot be null" },
  );

export type SourceUpdate = z.infer<typeof sourceUpdateSchema>;

export const sourceListQuerySchema = z.object({
  page: z.coerce.number().int().min(1).default(1),
  per_page: z.coerce.number().int().min(1).max(100).default(20),
  active_only: z
    .string()
    .transform((v) => v === "true")
    .default("false"),
});

export interface SourceResponse {
  id: string;
  name: string | null;
  rss_url: string;
  site_url: string | null;
  category: string | null;
  is_active: boolean;
  created_at: string;
}
